packages/client/src/index.ts

```diff
@@ -0,0 +1,378 @@
+    }
+
+    const { redirectUri, state, codeVerifier } = signInSession;
+    const code = verifyAndParseCodeFromCallbackUri(callbackUri, redirectUri, state);
+
+    const { appId: clientId } = logtoConfig;
+    const { tokenEndpoint } = await this.getOidcConfig();
+    const codeTokenResponse = await fetchTokenByAuthorizationCode(
+      {
+        clientId,
+        tokenEndpoint,
+        redirectUri,
+        codeVerifier,
+        code,
+      },
+      requester
+    );
+
+    await this.verifyIdToken(codeTokenResponse.idToken);
+
+    this.saveCodeToken(codeTokenResponse);
+    this.signInSession = null;
+  }
+
+  public async signOut(postLogoutRedirectUri?: string) {
+    if (!this.idToken) {
+      throw new LogtoClientError('not_authenticated');
+    }
+
+    const { appId: clientId } = this.logtoConfig;
+    const { endSessionEndpoint, revocationEndpoint } = await this.getOidcConfig();
+
+    if (this.refreshToken) {
+      try {
+        await revoke(revocationEndpoint, clientId, this.refreshToken, this.requester);
+      } catch {
+        // Do nothing at this point, as we don't want to break the sign-out flow even if the revocation is failed
+      }
+    }
+
+    const url = generateSignOutUri({
+      endSessionEndpoint,
+      postLogoutRedirectUri,
+      idToken: this.idToken,
+    });
+
+    this.accessTokenMap.clear();
+    this.refreshToken = null;
+    this.idToken = null;
+
+    this.handleRedirect(url);
+  }
+
+  private async getAccessTokenByRefreshToken(resource?: string): Promise<string> {
+    if (!this.refreshToken) {
+      throw new LogtoClientError('not_authenticated');
+    }
+
+    try {
+      const accessTokenKey = buildAccessTokenKey(resource);
+      const { appId: clientId } = this.logtoConfig;
+      const { tokenEndpoint } = await this.getOidcConfig();
+      const { accessToken, refreshToken, idToken, scope, expiresIn } =
+        await fetchTokenByRefreshToken(
+          {
+            clientId,
+            tokenEndpoint,
+            refreshToken: this.refreshToken,
+            resource,
+            scopes: resource ? ['offline_access'] : undefined, // Force remove openid scope from the request
+          },
+          this.requester
+        );
+
+      this.accessTokenMap.set(accessTokenKey, {
+        token: accessToken,
+        scope,
+        expiresAt: Math.round(Date.now() / 1000) + expiresIn,
+      });
+
+      this.refreshToken = refreshToken;
+
+      if (idToken) {
+        await this.verifyIdToken(idToken);
+        this.idToken = idToken;
+      }
+
+      return accessToken;
+    } catch (error: unknown) {
+      throw new LogtoClientError('get_access_token_by_refresh_token_failed', error);
+    }
+  }
+
+  private async _getOidcConfig() {
+    const { endpoint } = this.logtoConfig;
+    const discoveryEndpoint = getDiscoveryEndpoint(endpoint);
+
+    return fetchOidcConfig(discoveryEndpoint, this.requester);
+  }
+
+  private async _getJwtVerifyGetKey() {
```

Comment:
nit: JwtVerifyGetKey better name?

Reply:
follow the name from `jose`
<img width="2088" alt="image" src="https://user-images.githubusercontent.com/10594507/178403017-f89bf912-c232-4a7f-bb82-39e42bd859fd.png">